Remotion/Web/Development.WebTesting.IntegrationTests.Infrastructure/GenericTestParameterConverter.cs

```diff
@@ -1,69 +1,108 @@
+
+      if (reader.GetString() != "parameters")
+        throw CreateExceptionForExpectedTokenStringValue ("parameters", reader);
+
+      reader.Read();
+
+      var dictionary = JsonSerializer.Deserialize<Dictionary<string, IList<string>>> (ref reader);
+
+      reader.Read();
+
+      if (reader.TokenType != JsonTokenType.EndObject)
+        throw CreateExceptionForExpectedTokenType (JsonTokenType.EndObject, reader);
+
+      return new GenericTestPageParameterDto (
+          status,
+          dictionary?.ToDictionary (
+              kvp => kvp.Key,
+              kvp => new GenericTestPageParameter (kvp.Key, kvp.Value.ToArray()))
+          ?? new Dictionary<string, GenericTestPageParameter>());
+    }
+
+    public override void Write (Utf8JsonWriter writer, GenericTestPageParameterDto value, JsonSerializerOptions options)
     {
-      var status = (GenericTestPageStatus) (int) dictionary["status"];
-      var parameters = GetDeserializedParameterDictionary (dictionary["parameters"]);
+      writer.WriteStartObject();
+
+      writer.WriteNumber ("status", (int) value.Status);
+
+      writer.WriteStartObject ("parameters");
+      foreach (var parameter in value.Parameters)
+      {
+        writer.WritePropertyName (parameter.Key);
+        JsonSerializer.Serialize (writer, parameter.Value.Arguments);
+      }
+      writer.WriteEndObject();
 
-      return new GenericTestPageParameterDto (status, parameters);
+      writer.WriteEndObject();
     }
 
-    /// <inheritdoc />
-    public override IDictionary<string, object> Serialize (object obj, JavaScriptSerializer serializer)
+    public override bool CanConvert (Type typeToConvert)
     {
-      var information = (GenericTestPageParameterDto) obj;
-      return new Dictionary<string, object>
-             {
-                 { "status", (int) information.Status },
-                 { "parameters", information.Parameters.ToDictionary (pair => pair.Key, pair => pair.Value.Arguments) }
-             };
+      return typeToConvert == typeof (GenericTestPageParameterDto);
     }
 
-    private Dictionary<string, GenericTestPageParameter> GetDeserializedParameterDictionary (object serializedParameterDictionary)
+    private JsonException CreateExceptionForExpectedTokenType (JsonTokenType expected, Utf8JsonReader reader)
     {
-      return ((IDictionary<string, object>) serializedParameterDictionary)
-          .ToDictionary (pair => pair.Key, pair => new GenericTestPageParameter (pair.Key, CastToStringArray (pair.Value)));
+      return new ($"Expecting '{expected}', got '{reader.TokenType}' at index {reader.TokenStartIndex}.");
     }
 
-    private string[] CastToStringArray (object x)
+    private JsonException CreateExceptionForExpectedTokenStringValue (string expected, Utf8JsonReader reader)
     {
-      return ((ArrayList) x).Cast<string>().ToArray();
+      return new ($"Expecting value '{expected}', got '{reader.GetString()}' at index {reader.TokenStartIndex}.");
```

Comment:
Is this correct?
The docs for `reader.GetString()` state: _Reads the next JSON token value from the source, unescaped, and transcoded as a string._

So when you, for instance, make the `reader.GetString() != "status"` check in L45, throw the exception and call `GetString()` in `CreateExceptionForExpectedTokenStringValue` again - won't you get an incorrect value (the next token value) in `got '...'`?

Maybe I am misinterpreting this (I didn't execute it to check), but for such reasons, I personally would prefer not to pass `reader` into the exception construction methods, but only simple data - in order to reduce the risk of inducing side-effects (even though it might be necessary to store the result of a `reader`-invocation into a variable, then).

Reply:
That's an interesting point. Actually only `.Read()` advances the position in the stream, the various `.Get...` methods only extract the "next" (in fact current) value. So the output in the exception messages would be valid.

It still is a good point to only pass simple values to the exception factory methods. I'll leave this open for the lead review.

Reply:
Yes, I was not sure if the `.Get...` maybe could be doing an implicit read (and didn't look at the implementation, so I just asked to be sure). But if that's not the case it's nice :)